### ISSUE TITLE
cc_test coverage fails when fetching from remote cache

### ISSUE THREAD

**Issue by anhlinh123** [bazelbuild/bazel#20556]
### Description of the bug:

Running coverage on a cc_test randomly fails with the following error:
`I/O exception during sandboxed execution: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path_to_the_test_dir>/_coverage`
Afaik, the condition under which the bug happens might be:
- A coverage test passes and gets cached remotely.
- The artifacts of the last action (which collects coverage data) get evicted from the remote cache.
- The coverage test is triggered (in another job), and only the action is run, as everything else is already in the cache.
- The action tries to fetch the input artifacts, which contains a TreeArtifact (the _coverage directory).
- An exception is thrown https://github.com/bazelbuild/bazel/blob/7.0.0/src/main/java/com/google/devtools/build/lib/exec/SingleBuildFileCache.java#L78, probably because Bazel doesn't allow inputs to be directories.

### Which category does this issue belong to?

Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

- Use disk_cache instead of remote_cache to easily manipulate the cache. (add --disk_cache=<path_to_local_cache> to the command line).
- Run a cc_test: `bazel coverage <test> --disk_cache=<cache> --remote_upload_local_results=true --execution_log_json_file=<file_name>`.
- Open the json file, find the action that has `"commandArgs": ["external/bazel_tools/tools/test/collect_coverage.sh"],`
- Find the hash of the action output (near the end of the action object).
- Delete the file corresponding to the hash value in the cache directory.
- Rerun the test.
- The test fails with the error: `I/O exception during sandboxed execution: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path_to_the_test_dir>/_coverage`.

### Which operating system are you running Bazel on?

Ubuntu 20.04.6 LTS

### What is the output of `bazel info release`?

release 7.0.0

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

Probably it happens after this commit 1267631fee7523ea3ae572c9b8c9c1fc9c8c9452, when it tries to fetch metadata of a TreeArtifact (the _coverage directory).

### Have you found anything relevant by searching the web?

No.

**Comment by sputt** [comment]
Good writeup thank you.  This is blocking our upgrade to Bazel 7.0.0.  It also occurs with these flags set:
```
coverage --experimental_fetch_all_coverage_outputs
coverage --experimental_split_coverage_postprocessing
```

which we set due to requiring `nobuild_runfile_links`.

**Comment by tjgq** [comment]
@anhlinh123 @sputt If you run with `--verbose_failures`, are you able to get the stack trace leading up to the `I/O exception during sandboxed execution` exception? That would help narrow down the issue.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by fmeum** [comment]
@iancha1992 This looks like another 7.0.0 regression, just want to make sure it's tracked.

**Comment by iancha1992** [comment]
@fmeum @Wyverald @meteorcloudy Should this be included in the 7.0.1 patch? or is it okay to go straight to 7.1.0?

**Comment by Wyverald** [comment]
Since this is a regression in 7.0.0, ideally we should include a fix in 7.0.1. But that depends on the timeline. @tjgq do you have an estimate for how long the fix might take?

**Comment by iancha1992** [comment]
@bazel-io fork 7.0.1

**Comment by iancha1992** [comment]
@bazel-io fork 7.1.0

**Comment by iancha1992** [comment]
I'll include this in the 7.0.1 milestone anyway with the "nice-to-have" label then

**Comment by tjgq** [comment]
@nlou9 Same request as above: can you please run with `--verbose_failures` and post the full stack trace here?

**Comment by lberki** [comment]
Are you setting `--experimental_split_coverage_postprocessing` by any chance?

If so, I suspect that the root cause is the same as the coverage failure in #20753 and if so, have a fix for it that is in the process of being submitted (although CI is not really a happy camper now so it will take a bit of time to materialize at HEAD)

**Comment by lberki** [comment]
I *think* the idea is that 7.0.1 should out sometime next week the latest (don't take this as a promise until @meteorcloudy confirms)

**Comment by lberki** [comment]
Also, @tjgq suspects that there is *another* bug lurking in the deep and he was working on confirming or denying that. *My* guess is that the above change is enough, but that's just a guess and an uninformed one.

**Comment by tjgq** [comment]
If we have confirmation that the repro requires `--experimental_split_coverage_postprocessing`, I believe https://github.com/iancha1992/bazel/commit/9463a4f50981785da493097e1197799c525b6278 will fix it. (Otherwise, if it's supposed to repro without `--experimental_split_coverage_postprocessing`, I couldn't do so following the instructions above.)

**Comment by UebelAndre** [comment]
I see this on python tests as well. This issue blocks my ability to upgrade Bazel.

**Comment by anhlinh123** [comment]
@tjgq Sorry for the late reply. This is the error message when using --verbose_failures:
`ERROR: /<path_to_the_test>/BUILD:50:8: Testing //<path>:UnitTest failed: I/O exception during sandboxed execution: 11 errors during bulk transfer:
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage
com.google.devtools.build.lib.actions.DigestOfDirectoryException: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path>/UnitTest/_coverage`

**Comment by anhlinh123** [comment]
Ok, so it all started with this function: 
https://github.com/bazelbuild/bazel/blob/7.0.0/src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnCache.java#L82
which after all consumes all IOException:
https://github.com/bazelbuild/bazel/blob/7.0.0/src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnCache.java#L142

But the critical point seems to be here:
https://github.com/bazelbuild/bazel/blob/7.0.0/src/main/java/com/google/devtools/build/lib/remote/AbstractActionInputPrefetcher.java#L418
where it tries to fetch metadata of a tree artifact.
I'm not sure why it does that. But at that point, the tree artifact is a directory (I guess that is because the content of the directory was already evicted from the remote cache), and it breaks this code
https://github.com/bazelbuild/bazel/blob/7.0.0/src/main/java/com/google/devtools/build/lib/exec/SingleBuildFileCache.java#L78

**Comment by anhlinh123** [comment]
My repo also uses `--experimental_split_coverage_postprocessing`.
But I believe it doesn't need that flag to reproduce the error (if the flag is disabled by default).
It can be easily reproduce locally by this
```
What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.
Use disk_cache instead of remote_cache to easily manipulate the cache. (add --disk_cache=<path_to_local_cache> to the command line).
Run a cc_test: bazel coverage <test> --disk_cache=<cache> --remote_upload_local_results=true --execution_log_json_file=<file_name>.
Open the json file, find the action that has "commandArgs": ["external/bazel_tools/tools/test/collect_coverage.sh"],
Find the hash of the action output (near the end of the action object).
Delete the file corresponding to the hash value in the cache directory.
Rerun the test.
The test fails with the error: I/O exception during sandboxed execution: Input is a directory: bazel-out/k8-fastbuild/testlogs/<path_to_the_test_dir>/_coverage.
```

**Comment by tjgq** [comment]
@anhlinh123 Do you mind giving https://github.com/bazelbuild/bazel/commit/b0db044227d62178a7e578b8e03c452d8c17af33 a try? (The simplest way is to use Bazelisk with `USE_BAZEL_VERSION=latest`.) That would tell us whether there's indeed a separate issue that that commit didn't fix.

Otherwise, I think there might be something missing from your repro steps. You are building once, deleting a file from the disk cache, then rebuilding incrementally. I'd thus expect the incremental rebuild to be a no-op, since the output tree hasn't been touched and Bazel can tell it's up-to-date. Is the second build a clean build? Are there any other flags in a .bazelrc? (Use `--announce_rc` to print all of the flags Bazel is using.)

**Comment by tjgq** [comment]
(It's `USE_BAZEL_VERSION=last_green`, not `USE_BAZEL_VERSION=latest`. I've amended the previous comment.)

**Comment by anhlinh123** [comment]
@tjgq This is what I've found so far.
- I made a small project to reproduce the bug. You're right, I missed the step to clean up Bazel cache after tampering the remote cache.
- The flag `--experimental_split_coverage_postprocessing` actually matters. Turning it off did fix the error.
- Bazel last green fixes the error!!!

**Comment by tjgq** [comment]
That's great to hear, thanks. I will close this issue, since the fix has also been cherry-picked into the 7.0.1 branch (in https://github.com/bazelbuild/bazel/pull/20819).

**Comment by anhlinh123** [comment]
@tjgq thank you for your support!

**Comment by anhlinh123** [comment]
@tjgq The version 7.0.1rc1 has a weird bug related to this.
Turning `--experimental_split_coverage_postprocessing` on always breaks `coverage`.
```
ERROR: <path>/test/BUILD:1:8: Testing //:test failed: I/O exception during sandboxed execution: Input is a directory: bazel-out/k8-fastbuild/testlogs/test/_coverage
```
The `last_green` doesn't have it.

**Comment by tjgq** [comment]
7.0.0rc1 was cut before the fix made it into the 7.0.1 branch. Can you try `USE_BAZEL_VERSION=f4da34dcfe7b83388e3d963f35581a4fe710fc14` (current tip of the branch)?
